### ISSUE TITLE
Inconsistent dispose semantics

### ISSUE THREAD

**Issue by glopesdev** [dotnet/reactive#1110]
I recently ran into a consistency issue with call order of disposables. This is happening across all versions of Rx (2.2.5 > 4.0+). Consider the following queries:

```c#
static void Main(string[] args)
{
    var o1 = Observable.Using(
        () => Disposable.Create(() => Console.WriteLine("R1")),
        r => Observable.Return(0));

    var o2 = Observable.Using(
        () => Disposable.Create(() => Console.WriteLine("R2")),
        r => o1);

    o2.Subscribe();
    Console.ReadLine();
    Console.WriteLine("---");

    var o2a = o1.Publish(po1 => Observable.Using(
        () => Disposable.Create(() => Console.WriteLine("R2")),
        r => po1));

    o2a.Subscribe();
    Console.ReadLine();
    Console.WriteLine("---");

    var o2b = o1.SelectMany(x => Observable.Using(
        () => Disposable.Create(() => Console.WriteLine("R2")),
        r => Observable.Return(x)));

    o2b.Subscribe();
    Console.ReadLine();
    Console.WriteLine("---");
}
```

All of these queries are about chaining the results of a sequence as a causal consequence on the notifications of `o1`. Each of the observables also has some disposable resources that are specified with the `Using` operator.

The 1st example chains the two observables, by having `o2` subscribe to `o1` inside its creation expression. The 2nd example does exactly the same thing, but uses `Publish` to multicast the observable first. The 3rd example uses `SelectMany` to create `o2` from a sequence generated by the singleton value of `o1`.

All these observables return the exact same sequence of values. However, inconsistent behaviour is observed when cleaning up the resources involved in the operation.

Specifically the disposal order for the 1st example is:

```
R1
R2
```

Whereas the disposal order for the 2nd and 3rd examples is:

```
R2
R1
```

From being used to RAII principles, I would expect the latter case to be the correct one, as resources created last should be disposed first, which seems to be the case for the majority of operators in Rx.

Hence my surprise finding that the 1st simple example actually switches the order around. This is problematic for compositionality semantics of Rx. Any ideas about why this is happening and what should the expected semantics be like?

**Comment by akarnokd** [comment]
As far as I can tell (without debugging through), case 2 doesn't see the inner source's `IDisposable` in time so commences to dispose R2. Once the whole setup went through, `Publish` sees it has to dispose the inner source which then disposes R1. In other terms, `poi` runs to its completion synchronously but it's `IDisposable` isn't yet visible and is stepped over.

**Comment by glopesdev** [comment]
You are correct that the latest System.Reactive behaves like this only for the synchronous case.

Interestingly that is not the case for the old Rx 2.2.5: the disposal order is different regardless of whether the sequence is synchronous or asynchronous, so there is a confirmed difference in behaviour. I guess a lot of things have changed since then.

Is there a way to consistently swap the disposal order though? If you are composing operators which allocate resources that depend on previous resources, you would expect that dispose order is reversed, to ensure that the latest resources are disposed first, no? I'm finding it hard to understand how to think about chained resource disposal in the new Rx.

**Comment by glopesdev** [comment]
It's a bit concerning that in the new System.Reactive there are so many differences in semantics between the synchronous and asynchronous cases. I was used in the old Rx to treat synchronous and asynchronous sequences as interchangeable as far as subscription and disposal semantics were concerned, and I remember from the early videos that they were meant to be actually identical.

Is this a trend that is likely to continue?

**Comment by akarnokd** [comment]
It probably never worked but for a small fraction of cases via that trampolining logic in [`Producer`](https://github.com/dotnet/reactive/blob/master/Rx.NET/Source/src/System.Reactive/Internal/Producer.cs#L54). I think the usage of Rx.NET exploded lately and people are now running into this design problem now. In comparison, we went through similar issues with RxJava back in early 2014 which lead to a complete internal redesign.

**Comment by glopesdev** [comment]
My understanding from tracking the past few issues is that most of these inconsistencies have to do with the behaviour of default schedulers?

I'm sad for not being more aware of the internal implementation details of Rx, but my honest gut reaction is that at least by default the behaviour between synchronous and asynchronous cases should be the same. If there are opportunities for performance gains in some cases, then these should be specified explicitly, e.g. by using something like the old _Immediate_ scheduler.

**Comment by quinmars** [comment]
IIRC, this inconsitency (that the outer resource is disposed before the inner resource) slipped in with v4.1 and was fixed in v4.2. There are exist now unit tests to prevent such a regression for that case.

**Comment by quinmars** [comment]
Actually it was fixed in [v4.1.6](https://github.com/dotnet/reactive/releases/tag/rxnet-v4.1.6).

**Comment by bartdesmet** [comment]
This is indeed fixed, so closing this. Feel free to reactivate if you observe any inconsistency.
